Fix: allow to combine excludePackages and excludePackagesStartingWith options

eugene1g opened the pull request:
This PR makes it possible to combine the two different methods of excluding packages from evaluation (`excludePackages` and `excludePackagesStartingWith`).

Also fixes the incorrect behavior of `excludePackagesStartingWith` which resulted in the failed test at https://github.com/RSeidelsohn/license-checker-rseidelsohn/actions/runs/3924255813

Fixes #49

eugene1g commented:
@RSeidelsohn The recent [patch-release](https://github.com/RSeidelsohn/license-checker-rseidelsohn/compare/v3.2.0...v3.2.1#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R464) doesn't quite resolve the issue of combining those two attributes. I think this PR is still relevant.

eugene1g commented:
(rebased on the latest `master`)

RSeidelsohn commented:
> @RSeidelsohn The recent [patch-release](https://github.com/RSeidelsohn/license-checker-rseidelsohn/compare/v3.2.0...v3.2.1#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R464) doesn't quite resolve the issue of combining those two attributes. I think this PR is still relevant.

Definitely, @eugene1g , I unfortunately only saw your PR after I did my patch. I want to have a deeper look, though (or discuss here with you), as I did not check whether the shallow copy you do from `currentResult` into `resultJson` has some side effects. But I just had a quick look again and see that we only use the return value and should not have any issues with possible side effects.

But please - if this is ok - give me some more time (which might mean even days) to handle this. If it is urgent for you, tell me so and I will approve and merge your PR and make a new release. And thank you so much for contributing to this repo!

eugene1g commented:
@RSeidelsohn Sure thing - this is not urgent, and I'm happy to wait.

Refactoring `excludePackagesStartingWith` is not necessary for the fix to the issue. In fact, I probably should not have done it in the same PR. The reason why I did it is because it was doing both - creating a new object to return, and mutating values in the global `filtered` object, which is a no-no for predictability. I changed it to not have any side-effects, and refactored the loop to be more readable, but neither is _necessary_ for the main purpose of this PR, so you can restore the old function is you're more comfortable.

RSeidelsohn commented:
Thank you so much for your contribution! I really appreciate this, and you did an excellent work in also adding the tests (using a test framework I have never used before - I'd really like to switch to using Jest, but this would take quite some time).
I will prepare the new release now and also make it the last one guaranteed to be working with NodeJS v14, as [v14 will run out of security support in 3 months](https://endoflife.date/nodejs).
So the next major version will upgrade NodeJS to v18, which can be seen as a breaking change, which is why I will start a new major version.
